File: windows/DoNotType.App/TextInjector.cs
using DoNotType.Core;
using System.Runtime.InteropServices;
using System.Windows.Forms;

namespace DoNotType.App;

/// <summary>
/// Puts the transcript into whatever the user was typing in.
///
/// Clipboard save → Ctrl+V → restore. <c>SendInput</c> with per-character key events is the
/// alternative and behaves badly with anything non-ASCII, IME composition, or an app that
/// throttles input. The clipboard is borrowed, not taken: its contents are archived first and put
/// back afterwards.
/// </summary>
public static class TextInjector
{
    /// <summary>
    /// How long to let the target app read the clipboard before restoring it. Restoring too
    /// eagerly races the paste and inserts the user's old clipboard instead of the transcript.
    /// </summary>
    private static readonly TimeSpan RestoreDelay = TimeSpan.FromMilliseconds(250);

    public static async Task InsertAsync(string text, string dictation = "-")
    {
        if (string.IsNullOrEmpty(text)) return;

        // Logged because "it transcribed but nothing appeared" is a distinct failure from "it did
        // not transcribe", and from the outside they look the same. The window that receives the
        // keystrokes is the fact that separates them: a UAC-elevated window silently discards
        // input from an unelevated sender, and nothing anywhere reports that.
        var target = Interop.ForegroundWindowTitle();
        Log.Info(() => "inserting", new Dictionary<string, string>
        {
            ["dictation"] = dictation,
            ["chars"] = text.Length.ToString(),
            ["window"] = target.Length == 0 ? "untitled" : target,
        });

        var archive = SaveClipboard();
        SetClipboardText(text);

        SendPaste();
        await Task.Delay(RestoreDelay).ConfigureAwait(false);

        RestoreClipboard(archive);
        Log.Debug(
            () => "clipboard restored",
            new Dictionary<string, string> { ["dictation"] = dictation });
    }

    /// <summary>Puts the transcript on the clipboard and leaves it there.</summary>
    /// <remarks>
    /// For the cases where the paste must not happen: the words have been recorded, sent and paid
    /// for by then, and the difference between "press Ctrl+V" and "nothing happened" is the
    /// difference between one extra keystroke and an app that looks broken.
    /// </remarks>
    public static void CopyForManualPaste(string text, string dictation = "-")
    {
        if (string.IsNullOrEmpty(text)) return;
        SetClipboardText(text);
        Log.Info(() => "left on the clipboard for a manual paste", new Dictionary<string, string>
        {
            ["dictation"] = dictation,
            ["chars"] = text.Length.ToString(),
        });
    }

    /// <summary>Sends the configured submit keystroke after insertion and focus verification.</summary>
    public static bool Submit(FinishAndSendAction action, string dictation = "-")
    {
        if (action == FinishAndSendAction.Disabled) return false;
        var inputs = action == FinishAndSendAction.ModifiedEnter
            ? new[]
            {
                KeyEvent(Interop.VK_CONTROL, down: true),
                KeyEvent(Interop.VK_RETURN, down: true),
                KeyEvent(Interop.VK_RETURN, down: false),
                KeyEvent(Interop.VK_CONTROL, down: false),
            }
            : new[]
            {
                KeyEvent(Interop.VK_RETURN, down: true),
                KeyEvent(Interop.VK_RETURN, down: false),
            };
        var sent = Interop.SendInput(
            (uint)inputs.Length, inputs, Marshal.SizeOf<Interop.INPUT>()) == (uint)inputs.Length;
        if (sent)
        {
            Log.Info(() => "submission key sent", new Dictionary<string, string>
            {
                ["dictation"] = dictation,
                ["action"] = action.ToString(),
            });
        }
        else
        {
            Log.Warn(
                () => "submission key unavailable",
                new Dictionary<string, string> { ["dictation"] = dictation });
        }
        return sent;
    }

    private static readonly Log Log = new("inject");



    /// <summary>
    /// Every format currently on the clipboard, so a copied image or rich text survives.
    /// </summary>
    private static IDataObject? SaveClipboard()
    {
        try
        {
            var current = Clipboard.GetDataObject();
            if (current is null) return null;

            var copy = new DataObject();
            foreach (var format in current.GetFormats())
            {
                try
                {
                    var data = current.GetData(format);
                    if (data is not null) copy.SetData(format, data);
                }
                catch (ExternalException)
                {
                    // Some formats refuse to be read out of process; skip rather than abort.
                }
            }
            return copy;
        }
        catch (ExternalException)
        {
            // Another process holds the clipboard open. Proceed without an archive rather than
            // failing the dictation.
            return null;
        }
    }

    private static void RestoreClipboard(IDataObject? archive)
    {
        try
        {
            if (archive is null) Clipboard.Clear();
            else Clipboard.SetDataObject(archive, copy: true);
        }
        catch (ExternalException)
        {
            // Losing the previous clipboard is regrettable but not worth an error dialog.
        }
    }

    private static void SetClipboardText(string text)
    {
        for (var attempt = 0; attempt < 5; attempt++)
        {
            try
            {
                Clipboard.SetText(text);
                return;
            }
            catch (ExternalException)
            {
                // The clipboard is a single system-wide resource and is routinely held for a few
                // milliseconds by other apps; retrying is the documented remedy.
                Thread.Sleep(30);
            }
        }
    }

    /// <summary>
    /// Deletes the last <paramref name="count"/> characters with simulated backspaces.
    /// </summary>
    /// <remarks>
    /// By keystroke rather than by setting the field's value: the target window belongs to another
    /// application and most of them expose no settable value, which is the same reason insertion
    /// goes through the clipboard.
    ///
    /// Sent in one batch. A loop of individual SendInput calls interleaves with whatever the user
    /// types next, and the failure mode is deleting characters they typed after the dictation.
    /// </remarks>
    public static void DeleteBackward(int count, string dictation = "-")
    {
        if (count <= 0) return;

        Log.Info(() => "deleting an insertion", new Dictionary<string, string>
        {
            ["dictation"] = dictation,
            ["chars"] = count.ToString(),
            ["window"] = Interop.ForegroundWindowTitle(),
        });

        var inputs = new Interop.INPUT[count * 2];
        for (var index = 0; index < count; index++)
        {
            inputs[index * 2] = KeyEvent(Interop.VK_BACK, down: true);
            inputs[index * 2 + 1] = KeyEvent(Interop.VK_BACK, down: false);
        }
        Interop.SendInput((uint)inputs.Length, inputs, Marshal.SizeOf<Interop.INPUT>());
    }

    private static void SendPaste()
    {
        var inputs = new[]
        {
            KeyEvent(Interop.VK_CONTROL, down: true),
            KeyEvent(Interop.VK_V, down: true),
            KeyEvent(Interop.VK_V, down: false),
            KeyEvent(Interop.VK_CONTROL, down: false),
        };
        Interop.SendInput((uint)inputs.Length, inputs, Marshal.SizeOf<Interop.INPUT>());
    }

    private static Interop.INPUT KeyEvent(ushort virtualKey, bool down) => new()
    {
        type = Interop.INPUT_KEYBOARD,
        u = new Interop.InputUnion
        {
            ki = new Interop.KEYBDINPUT
            {
                wVk = virtualKey,
                dwFlags = down ? 0 : Interop.KEYEVENTF_KEYUP,
            },
        },
    };
}

File: Sources/DoNotTypeApp/TextInjector.swift
import AppKit
import ApplicationServices
import CoreGraphics
import DoNotTypeCore

/// Puts the transcript into whatever the user was typing in.
///
/// Pasteboard save → ⌘V → restore, which is what Typeless does and what every tool in this
/// category converges on. Setting `AXValue` directly is cleaner in principle and fails on too many
/// real apps to be the primary path — Electron, Java, terminals and anything canvas-drawn.
///
/// The clipboard is borrowed, not taken: every representation of every item is archived first and
/// put back afterwards, so a user's copied image or rich text survives a dictation.
@MainActor
enum TextInjector {
    private static let log = Log("inject")

    /// How long to let the target app read the pasteboard before restoring it.
    ///
    /// Restoring too eagerly races the paste and the user gets their old clipboard inserted
    /// instead of the transcript.
    private static let restoreDelay: Duration = .milliseconds(220)

    static func insert(_ text: String, dictation: String = "-") async {
        guard !text.isEmpty else { return }

        // Logged because "it transcribed but nothing appeared" is a distinct failure from "it did
        // not transcribe", and from the outside they look the same. Accessibility being off is the
        // usual cause and produces no error at all: the keystroke is simply never delivered.
        let trusted = AXIsProcessTrusted()
        let target = NSWorkspace.shared.frontmostApplication?.localizedName ?? "?"
        log.info(
            "inserting",
            [
                "dictation": dictation, "chars": "\(text.count)", "app": target,
                "accessibility": trusted ? "granted" : "MISSING",
            ])
        if !trusted {
            log.error(
                "cannot insert: Accessibility permission is not granted",
                ["dictation": dictation, "app": target])
        }

        let pasteboard = NSPasteboard.general
        let archive = archiveContents(of: pasteboard)

        pasteboard.clearContents()
        pasteboard.setString(text, forType: .string)

        simulatePaste()

        try? await Task.sleep(for: restoreDelay)
        restore(archive, to: pasteboard)
        log.debug("clipboard restored", ["dictation": dictation])
    }

    /// Puts the transcript on the clipboard and leaves it there.
    ///
    /// For the case where the paste cannot happen — Accessibility revoked, which macOS does on
    /// every signature change. The words have been recorded, sent and paid for by that point, and
    /// the difference between "press ⌘V" and "nothing happened" is the difference between a
    /// permission to fix and an app that looks broken.
    static func copyForManualPaste(_ text: String, dictation: String = "-") {
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(text, forType: .string)
        log.info(
            "left on the clipboard for a manual paste",
            ["dictation": dictation, "chars": "\(text.count)"])
    }

    // MARK: - Private

    /// Every type of every item, so non-text clipboard contents survive.
    private static func archiveContents(of pasteboard: NSPasteboard) -> [[NSPasteboard.PasteboardType: Data]] {
        (pasteboard.pasteboardItems ?? []).map { item in
            var representations: [NSPasteboard.PasteboardType: Data] = [:]
            for type in item.types {
                if let data = item.data(forType: type) { representations[type] = data }
            }
            return representations
        }
    }

    private static func restore(
        _ archive: [[NSPasteboard.PasteboardType: Data]], to pasteboard: NSPasteboard
    ) {
        pasteboard.clearContents()
        guard !archive.isEmpty else { return }

        let items = archive.map { representations -> NSPasteboardItem in
            let item = NSPasteboardItem()
            for (type, data) in representations { item.setData(data, forType: type) }
            return item
        }
        pasteboard.writeObjects(items)
    }

    /// Removes `count` characters before the caret.
    ///
    /// Simulated backspaces rather than setting the field's value: the target is an arbitrary app
    /// and most do not expose a settable value, which is the same reason insertion goes through
    /// the pasteboard. Batched with a small gap because some apps drop synthetic keys sent faster
    /// than a person could type.
    static func deleteBackward(count: Int) async {
        guard count > 0, let source = CGEventSource(stateID: .combinedSessionState) else { return }
        let backspace: CGKeyCode = 51

        for index in 0..<count {
            CGEvent(keyboardEventSource: source, virtualKey: backspace, keyDown: true)?
                .post(tap: .cgAnnotatedSessionEventTap)
            CGEvent(keyboardEventSource: source, virtualKey: backspace, keyDown: false)?
                .post(tap: .cgAnnotatedSessionEventTap)

            // Every few keystrokes, yield briefly. Without this a long transcript floods the
            // target app's event queue and some of the deletions are silently dropped.
            if index % 16 == 15 { try? await Task.sleep(for: .milliseconds(8)) }
        }
    }

    private static func simulatePaste() {
        guard let source = CGEventSource(stateID: .combinedSessionState) else {
            log.error("could not create an event source for ⌘V")
            return
        }
        // Suppress our own synthetic keystrokes from re-entering the app's own event tap.
        source.setLocalEventsFilterDuringSuppressionState(
            [.permitLocalMouseEvents, .permitSystemDefinedEvents], state: .eventSuppressionStateSuppressionInterval)

        let v: CGKeyCode = 9
        let down = CGEvent(keyboardEventSource: source, virtualKey: v, keyDown: true)
        let up = CGEvent(keyboardEventSource: source, virtualKey: v, keyDown: false)
        down?.flags = .maskCommand
        up?.flags = .maskCommand
        down?.post(tap: .cgAnnotatedSessionEventTap)
        up?.post(tap: .cgAnnotatedSessionEventTap)
    }
}
